File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if(nums1.length>nums2.length)
            return findMedianSortedArrays(nums2,nums1);
        int n1=nums1.length,n2=nums2.length,f=0,l=n1,mid=f+(l-f)/2;
        int req=(n1+n2)/2;
        if(((n1+n2)&1)!=0)
            ++req;
        while(f<=l){
            int f1=(mid-1)==-1?Integer.MIN_VALUE:nums1[mid-1];
            int f2=mid==n1?Integer.MAX_VALUE:nums1[mid];
            int idx=req-((mid-1)+1)-1;
            int s1=idx==-1?Integer.MIN_VALUE:nums2[idx];
            int s2=(idx+1)==n2?Integer.MAX_VALUE:nums2[idx+1];
            if(f1>s2){
                l=mid-1;
            }
            else if(s1>f2){
                f=mid+1;
            }
            else{
               // System.out.println(req);
                if(((n1+n2)&1)!=0)
                    return Math.max(f1,s1);
                double val1=Math.max(f1,s1);
                double val2=Math.min(f2,s2);
                // System.out.println(f2);
                // System.out.println(s1);
                // System.out.println(s2);
                return (val1+val2)/2;
            }
            mid=f+(l-f)/2;
        }
        return -1;
    }
}

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.kt
class Solution {
    fun findMedianSortedArrays(arr1: IntArray, arr2: IntArray): Double {
        if(arr2.size<arr1.size)return findMedianSortedArrays(arr2,arr1)
        var f:Int=0
        var l:Int=arr1.size
        var sz:Int=arr1.size+arr2.size
        var t:Int=sz/2
        while(f<=l){
            var mid:Int=f+((l-f) shr 1)
            var mid2:Int=t-mid
            if(mid2<0){
                l=mid-1
                continue
            }
            var l1:Int=Int.MIN_VALUE
            var l2:Int=Int.MIN_VALUE
            var r1:Int=Int.MAX_VALUE
            var r2:Int=Int.MAX_VALUE
            if(mid>0)l1=arr1[mid-1]
            if(mid2>0)l2=arr2[mid2-1]
            if(mid<arr1.size)r1=arr1[mid]
            if(mid2<arr2.size)r2=arr2[mid2]
            if(l2>r1){
                f=mid+1
            }else if(l1>r2){
                l=mid-1
            }
            else{
                if(sz and 1!=0)return min(r1,r2)*1.0
                return ((max(l1,l2)+min(r1,r2))*1.0)/2
            }
        }
        return -1.0
    }
}